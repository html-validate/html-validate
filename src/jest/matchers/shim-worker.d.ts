declare module "*?worker&url" {
	const value: string;
	export default value;
}
