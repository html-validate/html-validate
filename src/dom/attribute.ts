import { Location } from "../context";
import { DynamicValue } from "./dynamic-value";

interface StaticAttribute {
	readonly value: string | null;
}

interface DynamicAttribute {
	readonly value: DynamicValue;
}

/**
 * Type guard for [[Attribute]] testing if the value is static (missing value is
 * considered static as it is a fixed known value).
 *
 * @public
 */
export function isStaticAttribute(attr: Attribute | null): attr is Attribute & StaticAttribute {
	return Boolean(attr && attr.isStatic);
}

/**
 * Type guard for [[Attribute]] testing if the value is dynamic.
 *
 * @public
 */
export function isDynamicAttribute(attr: Attribute | null): attr is Attribute & DynamicAttribute {
	return Boolean(attr && attr.isDynamic);
}

/**
 * DOM Attribute.
 *
 * Represents a HTML attribute. Can contain either a fixed static value or a
 * placeholder for dynamic values (e.g. interpolated).
 */
export class Attribute {
	/** Attribute name */
	public readonly key: string;
	public readonly value: string | DynamicValue | null;
	public readonly keyLocation: Location;
	public readonly valueLocation: Location | null;
	public readonly originalAttribute?: string;

	/**
	 * @param key - Attribute name.
	 * @param value - Attribute value. Set to `null` for boolean attributes.
	 * @param keyLocation - Source location of attribute name.
	 * @param valueLocation - Source location of attribute value.
	 * @param originalAttribute - If this attribute was dynamically added via a
	 * transformation (e.g. vuejs `:id` generating the `id` attribute) this
	 * parameter should be set to the attribute name of the source attribute (`:id`).
	 */
	public constructor(
		key: string,
		value: null | string | DynamicValue | null,
		keyLocation: Location,
		valueLocation: Location | null,
		originalAttribute?: string
	) {
		this.key = key;
		this.value = value;
		this.keyLocation = keyLocation;
		this.valueLocation = valueLocation;
		this.originalAttribute = originalAttribute;

		/* force undefined to null */
		if (typeof this.value === "undefined") {
			this.value = null;
		}
	}

	/**
	 * Flag set to true if the attribute value is static.
	 */
	public get isStatic(): boolean {
		return !this.isDynamic;
	}

	/**
	 * Flag set to true if the attribute value is dynamic.
	 */
	public get isDynamic(): boolean {
		return this.value instanceof DynamicValue;
	}

	/**
	 * Test attribute value.
	 *
	 * @param pattern - Pattern to match value against. RegExp or a string (===)
	 * @param dynamicMatches - If true `DynamicValue` will always match, if false
	 * it never matches.
	 * @returns `true` if attribute value matches pattern.
	 */
	public valueMatches(pattern: RegExp, dynamicMatches?: boolean): boolean;
	public valueMatches(pattern: string, dynamicMatches?: boolean): boolean;
	public valueMatches(pattern: RegExp | string, dynamicMatches: boolean = true): boolean {
		if (this.value === null) {
			return false;
		}

		/* dynamic values matches everything */
		if (this.value instanceof DynamicValue) {
			return dynamicMatches;
		}

		/* test value against pattern */
		if (pattern instanceof RegExp) {
			return this.value.match(pattern) !== null;
		} else {
			return this.value === pattern;
		}
	}
}
