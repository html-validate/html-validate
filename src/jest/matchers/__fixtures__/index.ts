export { reportOk, reportOkAsync } from "./report-ok";
export { reportError, reportErrorAsync } from "./report-error";
export { reportMultipleErrors, reportMultipleErrorsAsync } from "./report-multiple-errors";
