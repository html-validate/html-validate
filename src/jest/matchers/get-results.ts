import { FileSystemConfigLoader } from "../../config/loaders/file-system";
import { type Report, type Result } from "../../reporter";
import HtmlValidate from "../../htmlvalidate";

/**
 * @internal
 */
export function getResults(filename: string, value: Report | string): Result[] {
	if (typeof value === "string") {
		const loader = new FileSystemConfigLoader({
			extends: ["html-validate:recommended"],
		});
		const htmlvalidate = new HtmlValidate(loader);
		const report = htmlvalidate.validateString(value, filename, {
			rules: {
				"void-style": "off",
			},
		});
		return report.results.map((it) => {
			return { ...it, filePath: "inline" };
		});
	} else {
		return value.results;
	}
}
