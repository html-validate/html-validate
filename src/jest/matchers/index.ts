export { toBeValid } from "./to-be-valid";
export { toBeInvalid } from "./to-be-invalid";
export { toHTMLValidate } from "./to-htmlvalidate";
export { toHaveError } from "./to-have-error";
export { toHaveErrors } from "./to-have-errors";
export { toMatchCodeframe } from "./to-match-codeframe";
export { toMatchInlineCodeframe } from "./to-match-inline-codeframe";
