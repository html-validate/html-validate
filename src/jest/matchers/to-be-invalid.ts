import { Report } from "../../reporter";
import { diverge } from "../utils";

function toBeInvalid(report: Report): jest.CustomMatcherResult {
	if (report.valid) {
		return {
			pass: false,
			message: () => "Result should be invalid but had no errors",
		};
	} else {
		return {
			pass: true,
			message: /* istanbul ignore next */ () => "Result should not contain error",
		};
	}
}

export default diverge(toBeInvalid);
