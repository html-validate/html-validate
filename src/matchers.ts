/* eslint-disable @typescript-eslint/no-namespace, @typescript-eslint/ban-ts-comment, prefer-template, sonarjs/no-duplicate-string */

import diff from "jest-diff";
import deepmerge from "deepmerge";
import { TokenType } from "./lexer";
import { Message, Report, Result } from "./reporter";
import { ConfigData } from "./config";
import HtmlValidate from "./htmlvalidate";

interface TokenMatcher {
	type: TokenType;
	location?: any;
	data?: any;
}

declare global {
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/ban-types */
		interface Matchers<R, T = {}> {
			toBeValid(): R;
			toBeInvalid(): R;
			toBeToken(expected: TokenMatcher): R;
			toHaveError(ruleId: string, message: string, context?: any): R;
			toHaveErrors(
				errors: Array<[string, string] | Record<string, unknown>>
			): R;

			/**
			 * Validate string or HTMLElement.
			 *
			 * Test passes if result is valid.
			 *
			 * @param config - Optional HTML-Validate configuration object.
			 */
			toHTMLValidate(config?: ConfigData): R;
		}
	}
}

/**
 * Takes all messages from all files and flattens to a single array.
 */
function flattenMessages(report: Report): Message[] {
	return report.results.reduce((aggregated: Message[], result: Result) => {
		return aggregated.concat(result.messages);
	}, []);
}

function toBeValid(
	this: jest.MatcherUtils,
	report: Report
): jest.CustomMatcherResult {
	if (report.valid) {
		return {
			pass: true,
			message: /* istanbul ignore next */ () =>
				"Result should not contain error",
		};
	} else {
		const firstError = report.results[0].messages[0];
		return {
			pass: false,
			message: () =>
				`Result should be successful but had error "${firstError.message}"`,
		};
	}
}

function toBeInvalid(
	this: jest.MatcherUtils,
	report: Report
): jest.CustomMatcherResult {
	if (report.valid) {
		return {
			pass: false,
			message: () => "Result should be invalid but had no errors",
		};
	} else {
		return {
			pass: true,
			message: /* istanbul ignore next */ () =>
				"Result should not contain error",
		};
	}
}

function toHaveError(
	this: jest.MatcherUtils,
	report: Report,
	ruleId: any,
	message: any,
	context?: any
): jest.CustomMatcherResult {
	const actual = flattenMessages(report);
	const expected: any = { ruleId, message };
	if (context) {
		expected.context = context;
	}

	const matcher = [expect.objectContaining(expected)];
	const pass = this.equals(actual, matcher);
	const diffString = diff(matcher, actual, { expand: this.expand });
	const resultMessage = (): string =>
		this.utils.matcherHint(".toHaveError") +
		"\n\n" +
		"Expected token to equal:\n" +
		`  ${this.utils.printExpected(matcher)}\n` +
		"Received:\n" +
		`  ${this.utils.printReceived(actual)}` +
		/* istanbul ignore next */ (diffString
			? `\n\nDifference:\n\n${diffString}`
			: "");

	return { pass, message: resultMessage };
}

function toHaveErrors(
	this: jest.MatcherUtils,
	report: Report,
	errors: Array<[string, string] | Record<string, unknown>>
): jest.CustomMatcherResult {
	const actual = flattenMessages(report);
	const matcher = errors.map((entry) => {
		if (Array.isArray(entry)) {
			const [ruleId, message] = entry;
			return expect.objectContaining({ ruleId, message });
		} else {
			return expect.objectContaining(entry);
		}
	});
	const pass = this.equals(actual, matcher);
	const diffString = diff(matcher, actual, { expand: this.expand });
	const resultMessage = (): string =>
		this.utils.matcherHint(".toHaveErrors") +
		"\n\n" +
		"Expected token to equal:\n" +
		`  ${this.utils.printExpected(matcher)}\n` +
		"Received:\n" +
		`  ${this.utils.printReceived(actual)}` +
		/* istanbul ignore next */ (diffString
			? `\n\nDifference:\n\n${diffString}`
			: "");

	return { pass, message: resultMessage };
}

function toHTMLValidate(
	this: jest.MatcherUtils,
	// @ts-ignore DOM library not available
	actual: string | HTMLElement,
	userConfig?: ConfigData,
	filename?: string
): jest.CustomMatcherResult {
	// @ts-ignore DOM library not available
	if (actual instanceof HTMLElement) {
		actual = actual.outerHTML;
	}

	const defaultConfig = {
		rules: {
			/* jsdom normalizes style so disabling rule when using this matcher or it
			 * gets quite noisy when configured with self-closing */
			"void-style": "off",
		},
	};
	const config = deepmerge(defaultConfig, userConfig || {});

	const actualFilename = filename || this.testPath;
	const htmlvalidate = new HtmlValidate();
	const report = htmlvalidate.validateString(actual, actualFilename, config);
	const pass = report.valid;
	if (pass) {
		return { pass, message: () => "HTML is valid when an error was expected" };
	} else {
		const errors = report.results[0].messages.map(
			(message) => `  ${message.message} [${message.ruleId}]`
		);
		return {
			pass,
			message: () =>
				["Expected HTML to be valid but had the following errors:", ""]
					.concat(errors)
					.join("\n"),
		};
	}
}

function toBeToken(
	this: jest.MatcherUtils,
	actual: any,
	expected: any
): jest.CustomMatcherResult {
	const token = actual.value;

	// istanbul ignore next: TokenMatcher requires "type" property to be set, this is just a failsafe
	if (token.type) {
		token.type = TokenType[token.type];
	}

	// istanbul ignore next: TokenMatcher requires "type" property to be set, this is just a failsafe
	if (expected.type) {
		expected.type = TokenType[expected.type];
	}

	const matcher = expect.objectContaining(expected);
	const pass = this.equals(token, matcher);
	const diffString = diff(matcher, token, { expand: this.expand });
	const message = (): string =>
		this.utils.matcherHint(".toBeToken") +
		"\n\n" +
		"Expected token to equal:\n" +
		`  ${this.utils.printExpected(matcher)}\n` +
		"Received:\n" +
		`  ${this.utils.printReceived(token)}` +
		/* istanbul ignore next */ (diffString
			? `\n\nDifference:\n\n${diffString}`
			: "");

	return { pass, message };
}

expect.extend({
	toBeValid,
	toBeInvalid,
	toHaveError,
	toHaveErrors,
	toHTMLValidate,
	toBeToken,
});
