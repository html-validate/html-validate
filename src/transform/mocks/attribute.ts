import { DynamicValue } from "../../dom";
import { type AttributeData } from "../../parser";

export function* processAttribute(attr: AttributeData): Iterable<AttributeData> {
	/* handle foo="{{ bar }}" as "foo" with a dynamic value (interpolated) */
	if (typeof attr.value === "string" && /{{.*}}/.exec(attr.value)) {
		yield {
			...attr,
			value: new DynamicValue(attr.value),
		};
		return;
	}

	/* passthru original attribute */
	yield attr;

	/* handle "dynamic-foo" as alias for "foo" with dynamic value */
	if (attr.key.startsWith("dynamic-")) {
		yield {
			...attr,
			key: attr.key.replace("dynamic-", ""),
			value: new DynamicValue(attr.value as string),
			originalAttribute: attr.key,
		};
	}
}
