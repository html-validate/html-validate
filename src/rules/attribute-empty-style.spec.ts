import HtmlValidate from "../htmlvalidate";
import "../matchers";
import { processAttribute } from "../transform/mocks/attribute";

describe("rule attribute-empty-style", () => {
	let htmlvalidate: HtmlValidate;

	it("should not report unknown elements", () => {
		expect.assertions(1);
		htmlvalidate = new HtmlValidate({
			rules: { "attribute-empty-style": ["error", { style: "omit" }] },
		});
		const report = htmlvalidate.validateString('<custom-element foobar=""></custom-element>');
		expect(report).toBeValid();
	});

	it("should not report unknown attributes", () => {
		expect.assertions(1);
		htmlvalidate = new HtmlValidate({
			rules: { "attribute-empty-style": ["error", { style: "omit" }] },
		});
		const report = htmlvalidate.validateString('<a foobar=""></a>');
		expect(report).toBeValid();
	});

	describe('configured with "omit"', () => {
		beforeAll(() => {
			htmlvalidate = new HtmlValidate({
				rules: { "attribute-empty-style": ["error", { style: "omit" }] },
			});
		});

		it("should report error when value is empty string", () => {
			expect.assertions(2);
			const report = htmlvalidate.validateString('<a download=""></a>');
			expect(report).toBeInvalid();
			expect(report).toHaveError("attribute-empty-style", 'Attribute "download" should omit value');
		});

		it("should not report error when value is omitted", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString("<a download></a>");
			expect(report).toBeValid();
		});

		it("should not report error for non-empty attributes", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString('<a download="file.txt"></a>');
			expect(report).toBeValid();
		});

		it("should not report error for boolean attributes", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString('<input required="">');
			expect(report).toBeValid();
		});

		it("should not report error when attribute is interpolated", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString('<a download="{{ dynamic }}">', {
				processAttribute,
			});
			expect(report).toBeValid();
		});

		it("should not report error when attribute is dynamic", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString('<input dynamic-required="dynamic">', {
				processAttribute,
			});
			expect(report).toBeValid();
		});
	});

	describe('configured with "empty"', () => {
		beforeAll(() => {
			htmlvalidate = new HtmlValidate({
				rules: { "attribute-empty-style": ["error", { style: "empty" }] },
			});
		});

		it("should report error when value is omitted", () => {
			expect.assertions(2);
			const report = htmlvalidate.validateString("<a download></a>");
			expect(report).toBeInvalid();
			expect(report).toHaveError(
				"attribute-empty-style",
				'Attribute "download" value should be empty string'
			);
		});

		it("should not report error when value is empty string", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString('<a download=""></a>');
			expect(report).toBeValid();
		});

		it("should not report error for non-empty attributes", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString('<a download="file.txt"></a>');
			expect(report).toBeValid();
		});

		it("should not report error for boolean attributes", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString('<input required="">');
			expect(report).toBeValid();
		});

		it("should not report error when attribute is interpolated", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString('<a download="{{ dynamic }}">', {
				processAttribute,
			});
			expect(report).toBeValid();
		});

		it("should not report error when attribute is dynamic", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString('<input dynamic-required="dynamic">', {
				processAttribute,
			});
			expect(report).toBeValid();
		});
	});

	it("should throw error if configured with invalid value", () => {
		expect.assertions(1);
		expect(() => {
			return new HtmlValidate({
				rules: { "attribute-empty-style": ["error", { style: "foobar" }] },
			});
		}).toThrowErrorMatchingInlineSnapshot(
			`"Rule configuration error: /rules/attribute-empty-style/1/style should be equal to one of the allowed values: empty, omit"`
		);
	});

	it("should contain documentation", () => {
		expect.assertions(1);
		htmlvalidate = new HtmlValidate({
			rules: { "attribute-empty-style": "error" },
		});
		expect(htmlvalidate.getRuleDocumentation("attribute-empty-style")).toMatchSnapshot();
	});
});
