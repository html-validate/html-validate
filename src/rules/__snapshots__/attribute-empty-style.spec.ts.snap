// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule attribute-empty-style should contain documentation 1`] = `
{
  "description": "Require a specific style for attributes with empty values.",
  "url": "https://html-validate.org/rules/attribute-empty-style.html",
}
`;
